### PR TITLE
Cosmos querying over arrays of structural type

### LINKED ISSUE #16926: Cosmos: Translate composition over embedded non-primitive collection
```C#
        [ConditionalFact]
        public virtual void Navigation_rewrite_on_owned_collection_with_composition()
        {
            using (var ctx = CreateContext())
            {
                var query = ctx.Set<OwnedPerson>().Select(p => p.Orders.Select(o => o.Id != 42).FirstOrDefault());
                var result = query.ToList();

                Assert.Equal(4, result.Count);
            }
        }
```

### LINKED ISSUE #20441: EFCore 3.1 + CosmosDb can't translate any() on owned object
I have a query that seems to be pretty basic. I'm using Core3.1 on top of CosmosDb. I simply want to query a list of orders that have order details, but I'm getting this error:

.Where(b => EF.Property>(b, "OrderDetails") .Any())' could not be translated. Either rewrite the query in a form that can be translated or switch to client evaluation.

Client evaluation isn't an option

Consider this contrived example:

```C#
public class Order
{
    public List<OrderDetail> OrderDetails {get;set;}
}

public class OrderDetail
{

}

// Db Context
modelBuilder.Entity<Order>().OwnsMany(x => x.OrderDetails);


// The query
context.Orders.Where( order => order.OrderDetails.Any());
```

I do this in my sleep with SQL.  What gives?

### LINKED ISSUE #25700: Cosmos: translate Length/Count on collections
```c#
_ = await ctx.Blogs.Where(b => b.Posts.Length == 3).ToListAsync();
```

Note that since we support primitive collections, this should translate on those too.

### LINKED ISSUE #25701: Cosmos: translate indexing into collection
```c#
_ = await ctx.Blogs.Where(b => b.Posts[2].Name == "foo").ToListAsync();
```

Note that since we support primitive collections, this should translate on those too.

### LINKED ISSUE #33032: Align Cosmos query pipeline code with latest relational code

### PR DESCRIPTION
**Pull request by roji**
This PR adds general support for querying over arrays of structural types in Cosmos. Most translations work, various problems and edge cases are tracked separately.

Note that this PR introduces a bit of expression type explosion, with many types representing the same NoSQL syntax (but returning different things, i.e. scalar vs. structural type). See #33999 for more information, hopefully we clean this up at some point.

Closes #16926
Closes #20441
Closes #25700
Closes #25701
Closes #33032

### CODE REVIEW

File: src/EFCore.Cosmos/Query/Internal/CosmosQueryUtils.cs [hunk 1]
```diff
@@ -53,12 +53,21 @@ public static class CosmosQueryUtils
...
-            array = new ArrayExpression(subquery, arrayClrType, arrayTypeMapping);
-            return true;
+            switch (projection)
+            {
+                case StructuralTypeShaperExpression:
+                    array = new ObjectArrayExpression(subquery, arrayClrType);
+                    return true;
+
+                // TODO: Temporary hack - need to perform proper derivation of the array type mapping from the element (e.g. for
+                // value conversion).
```
Comment (AndriySvyryd) [inline src/EFCore.Cosmos/Query/Internal/CosmosQueryUtils.cs:64]: Add an issue #

File: src/EFCore.Cosmos/Query/Internal/CosmosQueryableMethodTranslatingExpressionVisitor.cs [hunk 2]
```diff
@@ -531,28 +578,57 @@ protected override ShapedQueryExpression TranslateDistinct(ShapedQueryExpression
...
 
-            return source.UpdateQueryExpression(new SelectExpression(translation));
+            // ElementAtOrDefault over an array os structural types
+            case not null when projectedStructuralTypeShaper is not null:
+            {
+                var translation = new ObjectArrayIndexExpression(array, translatedIndex, projectedStructuralTypeShaper.Type);
+
+                if (returnDefault)
+                {
+                    // TODO
```
Comment (AndriySvyryd) [inline src/EFCore.Cosmos/Query/Internal/CosmosQueryableMethodTranslatingExpressionVisitor.cs:616]: 👀

File: src/EFCore.Cosmos/Query/Internal/CosmosQueryUtils.cs (hunk 1 above)
Comment (roji) [reply src/EFCore.Cosmos/Query/Internal/CosmosQueryUtils.cs:64]: Done, #34026

File: src/EFCore.Cosmos/Query/Internal/CosmosQueryableMethodTranslatingExpressionVisitor.cs (hunk 2 above)
Comment (roji) [reply src/EFCore.Cosmos/Query/Internal/CosmosQueryableMethodTranslatingExpressionVisitor.cs]: Thanks, implemented.

**Review by AndriySvyryd — APPROVED** [review]

**Pull request merged**